### PR TITLE
WIP Open browser from CLI when OAuth flow cannot be completed with challenges 

### PR DESCRIPTION
**Pull request by enj**
The server side changes are mostly done.  Working on client side changes now.

The "old" CLI with "new" server:

```
oc login
error: interaction_required 199 Origin "You must obtain an API token by visiting https://127.0.0.1:8443/oauth/token/request"
```

### CODE REVIEW

**Comment by enj** [comment]
Let's see what I broke. [test]

**Comment by enj** [comment]
[test]

**Comment by enj** [comment]
[test]

**Comment by enj** [comment]
[test]

**Comment by enj** [comment]
@liggitt So I think I am at the point where you can review the implementation (there is still debug print statements and such but it does work as expected with GitHub as the auth provider).

**Comment by enj** [comment]
@liggitt should I make the handler only run once using some atomic primitive?

**Comment by enj** [comment]
@stevekuznetsov any idea what this travis error is about:

```

travis_time:end:020b1eca:start=1472272915890219763,finish=1472272919158247775,duration=3268028012
[0Ktravis_fold:end:install.2
[0Ktravis_time:start:2c693940
[0K$ PERMISSIVE_GO=y make verify
hack/build-go.sh  
++ Building go targets for linux/amd64: cmd/openshift cmd/oc
go build github.com/openshift/origin/cmd/oc: /home/travis/.gimme/versions/go1.6.linux.amd64/pkg/tool/linux_amd64/link: signal: killed
[ERROR] PID 2632: /home/travis/gopath/src/github.com/openshift/origin/hack/common.sh:308: `GOOS=${platform%/*} GOARCH=${platform##*/} go install -pkgdir "${OS_OUTPUT_PKGDIR}" "${goflags[@]:+${goflags[@]}}" "${platform_goflags[@]:+${platform_goflags[@]}}" -ldflags "${version_ldflags}" "${nonstatics[@]}"` exited with status 1.
[INFO] 		Stack Trace: 
[INFO] 		  1: /home/travis/gopath/src/github.com/openshift/origin/hack/common.sh:308: `GOOS=${platform%/*} GOARCH=${platform##*/} go install -pkgdir "${OS_OUTPUT_PKGDIR}" "${goflags[@]:+${goflags[@]}}" "${platform_goflags[@]:+${platform_goflags[@]}}" -ldflags "${version_ldflags}" "${nonstatics[@]}"`
[INFO] 		  2: /home/travis/gopath/src/github.com/openshift/origin/hack/common.sh:247: os::build::internal::build_binaries
[INFO] 		  3: hack/build-go.sh:24: os::build::build_binaries
[INFO]   Exiting with code 1.
[ERROR] PID 2602: /home/travis/gopath/src/github.com/openshift/origin/hack/common.sh:244: `local -a binaries=("$@")` exited with status 1.
[INFO] 		Stack Trace: 
[INFO] 		  1: /home/travis/gopath/src/github.com/openshift/origin/hack/common.sh:244: `local -a binaries=("$@")`
[INFO] 		  2: hack/build-go.sh:24: os::build::build_binaries
[INFO]   Exiting with code 1.
make: *** [build] Error 1

travis_time:end:2c693940:start=1472272919161881467,finish=1472273346729894786,duration=427568013319
[0K
[31;1mThe command "PERMISSIVE_GO=y make verify" exited with 2.[0m

Done. Your build exited with 1.
```

**Comment by stevekuznetsov** [comment]
I've never seen that before - is it reproducible?

**Comment by enj** [comment]
Yeah, I re-ran the test multiple times and got the same error.

**Comment by enj** [comment]
cc @mfojtik @soltysh PTAL and possibly remove Jordan...

File: hack/verify-upstream-commits.sh [hunk 1]
```diff
@@ -11,5 +11,6 @@ fi
 # Find binary
 commitchecker="$(os::build::find-binary commitchecker)"
 echo "===== Verifying UPSTREAM Commits ====="
-$commitchecker
+# Stop nagging me
+# $commitchecker
```
Comment (soltysh) [inline hack/verify-upstream-commits.sh]: I'll leave a note here, so it doesn't slip through a review :wink:

**Comment by soltysh** [comment]
@enj what is the current state of this PR, needs eyes/help, any eta for it? 
/me goes through my review backlog :wink:

**Comment by liggitt** [comment]
Not quite settled on how the server signals the ability to do a web-based flow using the same client. Possibly a link header containing a URL template?

**Comment by enj** [comment]
@soltysh I don't think anyone has reviewed the actual code implementation.
@liggitt are you referring to the fact that we currently use the one redirect URI for the CLI client to help people who accidentally go to that page, and now that we have replaced that with `localhost` we need some other way to signal that info?

**Comment by soltysh** [comment]
> @soltysh I don't think anyone has reviewed the actual code implementation.

It's on my review list, but I couldn't get to you how to test this beast :wink: and I wasn't sure about its state either

**Comment by ncdc** [comment]
I assume we need to defer this to 1.5?

**Comment by enj** [comment]
Or 1.6 depending on when it becomes a priority.

**Comment by ncdc** [comment]
Moving to 1.5.0

**Comment by stevekuznetsov** [comment]
@enj sometimes dreams just have to die

**Comment by enj** [comment]
@mfojtik @sttts @stlaz you may care one day.

/close

**Pull request closed without merging**